Reading Docker timeout value from environment variable

mael-le-gal opened the pull request:
Fixing the following [issue](https://github.com/jupyter/repo2docker/issues/711) by adding `DOCKER_REGISTRY_TIMEOUT` environment variable

consideRatio commented:
Hi @mael-le-gal I mistakenly cancelled your CI build and restarted it again! Sorry!

I noted that one job of all the jobs had failed relating to not having run the autoformatter `black`.

consideRatio commented:
My understanding is that this PR makes it so that repo2docker configures docker to explicitly use its own default value, unless a environment variable is set, called `DOCKER_REGISTRY_TIMEOUT`.

Considerations:
- I wonder, would docker adjust itself if we simply provided an env var called `DEFAULT_TIMEOUT_SECONDS` ?
- I wonder, if we add a new env that has meaning to repo2docker, should we perhaps prefix it with `REPO2DOCKER` or `R2D`? I've come to appreciate such things a lot as it can be quite a hurdle to figure out why and who considers what env vars if you have something nesting something that potentially even nests something else etc.

consideRatio commented:
I've learned that no, Docker does not care about an environment variable named `DOCKER_TIMEOUT_SECONDS`.

__Hardcoded constant...__
https://github.com/docker/docker-py/blob/20b5491363a9124dbf53b00fb6562f8f257a6228/docker/constants.py#L1-L6

__... used as fallback if no timeout argument was provided.__
https://github.com/docker/docker-py/blob/20b5491363a9124dbf53b00fb6562f8f257a6228/docker/client.py#L82

mael-le-gal commented:
> I noted that one job of all the jobs had failed relating to not having run the autoformatter black.

Indeed, it should be fixed now

> I wonder, would docker adjust itself if we simply provided an env var called DEFAULT_TIMEOUT_SECONDS ?

No there is only a hardcoded default timeout but it is never override by environment variable

> Regarding my second consideration, I'm leaning towards naming the variable REPO2DOCKER_DOCKER_TIMEOUT or something making it obviously ours.

Fixed, I changed to `REPO2DOCKER_DOCKER_TIMEOUT`

consideRatio commented:
Should we set the timeout explicitly, by using the docker default value imported from the docker package at? It will entwine us with the docker package slightly. If we need that, because we need to know the timeout someplace for some reason, then I think we should. If we don't need to know the implicitly set timeout, then I think it could be beneficial to avoid explicitly passing a `timeout` argument when we don't detect a `REPO2DOCKER_DOCKER_TIMEOUT` variable.

betatim commented:
Two general comments:

* we use https://github.com/ipython/traitlets for the configuration of repo2docker, lets use it also for this timeout. I need to dig a bit to find an example/docs of how traitlets and env variables work together. We should use that machinery as much as we can
* not sure how I feel about fetching the default value and using it. I'd have used `None` as a value for the "default" value in r2d which signals use what ever the default in docker-py is. However [docker-py sets the default](https://github.com/docker/docker-py/blob/4d8e5eb2ada3dfbffd5721271c1c7c6bec476dd4/docker/api/client.py#L102-L105) to a particular value (not `None`) so I think in this case getting the value and passing it in is the thing to do.

betatim commented:
I'd suggest that we close this PR. It hasn't been a problem in a long time and we can open a new PR using traitlets to implement this if we need to.

(We are trying to reduce the number of outstanding/stale PRs)
